tests/functional/test-common.h

```diff
@@ -205,7 +265,499 @@ static inline ncclResult_t validate_data(char *recv_buf, char *expected_buf, siz
+			if (done) {
+				completed[i] = true;
+				num_completed++;
+				NCCL_OFI_TRACE(NCCL_NET, "Request %zu completed (%zu/%zu)", i, num_completed, num_requests);
+			}
+		}
+
+		// Check if all requests are complete
+		if (num_completed >= num_requests) {
+			NCCL_OFI_TRACE(NCCL_NET, "All %zu requests completed", num_requests);
+			return ncclSuccess;
+		}
+
+		// Check if we need to timeout
+		if (timeout_ms == 0) {
+			std::this_thread::sleep_for(std::chrono::milliseconds(1));
+			continue;
+		}
+		const auto current_time = std::chrono::steady_clock::now();
+		const auto elapsed = current_time - start_time;
+		if (elapsed >= timeout_duration) {
+			NCCL_OFI_WARN("Timeout: %zu/%zu requests completed",
+		 num_completed, num_requests);
+			return ncclInternalError;
+		}
+	}
+
+	// We shouldn't be getting here
+	return ncclInternalError;
+}
+
+/**
+ * Setup connection between two ranks
+ *
+ * This helper establishes a bidirectional connection between the current rank
+ * and a peer rank. It creates a listen communicator, exchanges connection handles
+ * via MPI, and creates send and receive communicators.
+ *
+ * @param ext_net Pointer to external plugin
+ * @param dev Device index to use for connection
+ * @param rank Current MPI rank
+ * @param size Total number of MPI ranks
+ * @param peer_rank MPI rank of the peer to connect to
+ * @param ndev Number of devices
+ * @param lcomm Output: listen communicator (may be NULL after return)
+ * @param scomm Output: send communicator
+ * @param rcomm Output: receive communicator
+ * @param sHandle Output: send device handle
+ * @param rHandle Output: receive device handle
+ * @return ncclSuccess on success, error code otherwise
+ */
+inline ncclResult_t setup_connection(
+	test_nccl_net_t* ext_net,
+	int dev,
+	int rank,
+	int size,
+	int peer_rank,
+	int ndev,
+	int tag,
+	nccl_net_ofi_listen_comm_t** lcomm,
+	nccl_net_ofi_send_comm_t** scomm,
+	nccl_net_ofi_recv_comm_t** rcomm,
+	test_nccl_net_device_handle_t** shandle,
+	test_nccl_net_device_handle_t** rhandle)
+{
+	// Validate device index
+	if (dev < 0 || dev >= ndev) {
+		NCCL_OFI_WARN("Invalid device index %d (ndev=%d)", dev, ndev);
+		return ncclInvalidArgument;
+	}
+
+	// Validate peer rank
+	if (peer_rank < 0 || peer_rank >= size || peer_rank == rank) {
+		NCCL_OFI_WARN("Invalid peer rank %d (rank=%d, size=%d)",
+		peer_rank, rank, size);
+		return ncclInvalidArgument;
+	}
+
+	NCCL_OFI_TRACE(NCCL_INIT, "Rank %d setting up connection with rank %d on device %d",
+		rank, peer_rank, dev);
+
+	// Initialize output pointers
+	*lcomm = nullptr;
+	*scomm = nullptr;
+	*rcomm = nullptr;
+	*shandle = nullptr;
+	*rhandle = nullptr;
+	char local_handle[NCCL_NET_HANDLE_MAXSIZE] = {};
+	char peer_handle[NCCL_NET_HANDLE_MAXSIZE] = {};
+
+
+
+	// Create listen communicator
+	NCCL_OFI_TRACE(NCCL_INIT, "Rank %d: Creating listen communicator on device %d",
+		rank, dev);
+	OFINCCLCHECK(ext_net->listen(dev, static_cast<void*>(&local_handle),
```

Comment:
Are you going to add a new scenario/function for testing new listen and connect API for requesting multiple domains?

Reply:
we also need a scenario to test the max limit for number of connections.